Adding docs for using an External Keycloak

davidfestal opened the pull request:
### What does this PR do?

This PR adds documentation about configuring Che to work with an external Keycloak instance.  

### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/17358

### Specify the version of the product this PR applies to.

7.18.0

### Still to be done

- provide specific links to RHSSO docs, instead of the upstream Keycloak docs, in the downstream case.

nickboldt commented on src/main/pages/che-7/administration-guide/proc_configuring-prod-short-to-use-external-keycloak.adoc:
```diff
@@ -0,0 +1,77 @@
+====
+* {prod-short} only supports `public` OIDC clients. Therefore, selecting the `openid-connect` `Client Protocol` option and the `public` Access Type option is highly recommended.
+
+* The list of `Valid Redirect URIs` must contain at least 2 URIs related to the {prod-short} server, one using the `http` protocol and the other `https`. These URIs must contain the base URL of the {prod-short} server, followed by `/*` wildcards.
```
do we really need http? I thought we were trying to do away with insecure http...

davidfestal replied:
Well, yes, we could drop http and keep https for the default case, but having it is just a way to manage all the cases (even non-default ones).
BTW I just added in the docs what is done by the operator itself in case of internal Keycloak.

But if you have a strong opinion about that I assume we can remove the `http` one.

nickboldt commented on src/main/pages/che-7/administration-guide/proc_configuring-prod-short-to-use-external-keycloak.adoc:
```diff
@@ -0,0 +1,77 @@
+
+* The list of `Valid Redirect URIs` must contain at least 2 URIs related to the {prod-short} server, one using the `http` protocol and the other `https`. These URIs must contain the base URL of the {prod-short} server, followed by `/*` wildcards.
+
+* The list of `Web Origins` must contain at least 2 URIs related to the Che server, one using the `http` protocol and the other `https`. These URIs must contain the base URL of the {prod-short} server, without any path after the host.
```
same question here ... given CRW 2.3 requires TLS, is there even a way to refer to the che server route by http url?

davidfestal replied:
Well, yes, we could drop http and keep https for the default case, but having it is just a way to manage all the cases (even non-default ones).
BTW I just added in the docs what is done by the operator itself in case of internal Keycloak.

But if you have a strong opinion about that I assume we can remove the `http` one.

nickboldt commented on src/main/pages/che-7/administration-guide/proc_configuring-prod-short-to-use-external-keycloak.adoc:
```diff
@@ -0,0 +1,77 @@
+
+* The list of `Web Origins` must contain at least 2 URIs related to the Che server, one using the `http` protocol and the other `https`. These URIs must contain the base URL of the {prod-short} server, without any path after the host.
++
+The number of URIs depends on the number of installed product tools.
```
what are product tools?

davidfestal replied:
The idea behind this this sentence was the following:

> it might be that some other URIs are already  configured if this public client is also used to access Keycloak from other company tools.

nickboldt commented on src/main/pages/che-7/administration-guide/proc_configuring-prod-short-to-use-external-keycloak.adoc:
```diff
@@ -0,0 +1,77 @@
+endif::[]
+and uses the default OpenShift OAuth support, user authentication relies on the integration of {identity-provider} with OpenShift OAuth. This allows users to log in to {prod-short} with their OpenShift login and have their workspaces created under personal OpenShift namespaces. 
++
+This requires setting up an OpenShift identity provider inside {identity-provider}. When using an external {identity-provider}, set up the identity provider manually. For instructions, seethe appropriate {identity-provider} documentations for either link:https://www.keycloak.org/docs/latest/server_admin/#openshift-3[OpenShift 3] or link:https://www.keycloak.org/docs/latest/server_admin/#openshift ...
```
typo: `seethe ` :D Unless you mean you're super duper angry :P

nickboldt replied:
in DOWNSTREAM only, should we link to the SSO docs instead of the Keycloak ones?

MichalMaler replied:
well, it will look exactly the same, but sure.

davidfestal replied:
Yes, that's the only thing I've not done in this PR: we should use attributes here to manage different URL for upstream and downstream.
@MichalMaler Do you plan to do it or should I do it myself ?

MichalMaler replied:
Will do it.

MichalMaler replied:
Here is the Downstream MR that will add the CRW identity provider links to the downstream docs
https://gitlab.cee.redhat.com/red-hat-developers-documentation/red-hat-devtools/-/merge_requests/715

davidfestal commented on src/main/pages/che-7/administration-guide/proc_configuring-prod-short-to-use-external-keycloak.adoc:
```diff
@@ -3,44 +3,44 @@
 = Configuring {prod-short} to use an external {identity-provider} installation
 
-By default, {prod-short} installation in multiuser mode includes the deployment of a dedicated {identity-provider} instance. But using an external {identity-provider} is also possible.
+By default, {prod-short} installation in multiuser mode includes the deployment of a dedicated {identity-provider} instance. However, using an external {identity-provider} is also possible. This option is useful when a user has an already existing {identity-provider} instance with already defined users.
```
```suggestion
By default, {prod-short} installation in multiuser mode includes the deployment of a dedicated {identity-provider} instance. However, using an external {identity-provider} is also possible. This option is useful when a user has an already existing {identity-provider} instance with already defined users, for example, a company-wide {identity-provider} server used by several applications.
```

davidfestal commented on src/main/pages/che-7/administration-guide/proc_configuring-prod-short-to-use-external-keycloak.adoc:
```diff
@@ -3,44 +3,44 @@
 
-*  In the administration console of the {identity-provider} external installation, you shoud have defined a link:https://www.keycloak.org/docs/latest/server_admin/#_create-realm[realm] that will contain the users who should connect to {prod-short}. Let's call it `realm-for-users` in the following, but in the next steps you would replace this name by the real one.
-
+*  In the administration console of the {identity-provider} external installation, define a link:https://www.keycloak.org/docs/latest/server_admin/#_create-realm[realm] that will contain the users intended to connect to {prod-short}. The following steps use the example `realm-for-users` realm name.
```
In fact maybe this realm already exist and has already be defined (for example in a company-wide Keycoak instance), but can be reused for Che.

So we should either:
- identify an already-existing that contains the users we're interested in
- Or create one if there's none.

MichalMaler replied:
Any steps that could tell that the realm already exists and could be used for Che @davidfestal ?

davidfestal commented on src/main/pages/che-7/administration-guide/proc_configuring-prod-short-to-use-external-keycloak.adoc:
```diff
@@ -3,44 +3,44 @@
+This requires setting up an OpenShift identity provider inside {identity-provider}. When using an external {identity-provider}, set up the identity provider manually. For instructions, seethe appropriate {identity-provider} documentations for either link:https://www.keycloak.org/docs/latest/server_admin/#openshift-3[OpenShift 3] or link:https://www.keycloak.org/docs/latest/server_admin/#openshift ...
 
-CAUTION: In the configured identity provider you should have the *Store Tokens* and *Stored Tokens Readable* enabled
+* The configured identity provider has `Store Tokens` and `Stored Tokens Readable` options enabled.
```
That's not a new bullet point, it's a warning in the current point no ?

MichalMaler replied:
Since the original was: `CAUTION: In the configured identity provider you should have the *Store Tokens* and *Stored Tokens Readable* enabled`, I created a Prerequsitiy bullet point out of it, since the sentence is saying that it needs to be enabled.

davidfestal commented on src/main/pages/che-7/administration-guide/proc_configuring-prod-short-to-use-external-keycloak.adoc:
```diff
@@ -38,7 +38,7 @@ on OpenShift
 and uses the default OpenShift OAuth support, user authentication relies on the integration of {identity-provider} with OpenShift OAuth. This allows users to log in to {prod-short} with their OpenShift login and have their workspaces created under personal OpenShift namespaces. 
 +
-This requires setting up an OpenShift identity provider inside {identity-provider}. When using an external {identity-provider}, set up the identity provider manually. For instructions, seethe appropriate {identity-provider} documentations for either link:https://www.keycloak.org/docs/latest/server_admin/#openshift-3[OpenShift 3] or link:https://www.keycloak.org/docs/latest/server_admin/#openshift ...
+This requires setting up an OpenShift identity provider inside {identity-provider}. When using an external {identity-provider}, set up the identity provider manually. For instructions, seethe appropriate {identity-provider} documentations for either link:https://access.redhat.com/documentation/en-us/red_hat_single_sign-on/7.4/html/server_administration_guide/identity_broker#openshift_3[OpenShift  ...
```
@MichalMaler We should have both URLs here, and have conditionals that drive whether we put either the upstream or the downstream URLs

amisevsk replied:
The correct way to fix this problem is to add keys to `_config.yml` as in https://github.com/eclipse/che-docs/pull/1423

amisevsk commented on src/main/pages/che-7/administration-guide/proc_configuring-prod-short-to-use-external-keycloak.adoc:
```diff
@@ -0,0 +1,77 @@
+[id="configuring-prod-short-to-use-external-keycloak_{context}"]
+= Configuring {prod-short} to use an external {identity-provider} installation
+
+By default, {prod-short} installation in multiuser mode includes the deployment of a dedicated {identity-provider} instance. However, using an external {identity-provider} is also possible. This option is useful when a user has an already existing {identity-provider} instance with already defined users.
```
Redundant word + add hyphen (already-defined vs already defined)
```suggestion
By default, {prod-short} installation in multiuser mode includes the deployment of a dedicated {identity-provider} instance. However, using an external {identity-provider} is also possible. This option is useful when a user has an existing {identity-provider} instance with already-defined users.
```

amisevsk commented on src/main/pages/che-7/administration-guide/proc_configuring-prod-short-to-use-external-keycloak.adoc:
```diff
@@ -0,0 +1,77 @@
+.Prerequisites
+
+ifeval::["{project-context}" == "che"]
+* This procedure is only applicable to {prod-short} installations done using the {prod-short} Operator. When using the `{prod-cli}` management tool and the `helm` installation mode to install {prod-short}, no supported method is available to use an external {identity-provider} instance.  
```
This is unclear -- is it not supported when using `chectl` or when using `chectl` + the helm installation mode? If it's the former, we can drop "and the `helm` installation mode"

amisevsk commented on src/main/pages/che-7/administration-guide/proc_configuring-prod-short-to-use-external-keycloak.adoc:
```diff
@@ -0,0 +1,77 @@
++
+This requires setting up an OpenShift identity provider inside {identity-provider}. When using an external {identity-provider}, set up the identity provider manually. For instructions, seethe appropriate {identity-provider} documentations for either link:https://access.redhat.com/documentation/en-us/red_hat_single_sign-on/7.4/html/server_administration_guide/identity_broker#openshift_3[OpenShift  ...
+
+* The configured identity provider has `Store Tokens` and `Stored Tokens Readable` options enabled.
```
```suggestion
* The configured identity provider has the options `Store Tokens` and `Stored Tokens Readable` enabled.
```

amisevsk commented on src/main/pages/che-7/administration-guide/proc_configuring-prod-short-to-use-external-keycloak.adoc:
```diff
@@ -0,0 +1,77 @@
+    # CAUTION: realm-for-users should be replaced by the name of your realm
+    identityProviderRealm: realm-for-users
+    # CAUTION: public-client should be replaced by the name of your client
+    identityProviderClientId: public-client
```
Our standard I've seen is to define a placeholder at the top, state what it is, and then use it throughout, e.g. `<auth-realm>` and `<client-id>`

rkratky commented on src/main/pages/che-7/administration-guide/assembly_configuring-authorization.adoc:
```diff
@@ -23,6 +23,7 @@ include::proc_setting-up-user-federation.adoc[leveloffset=+1]
 include::proc_enabling-authentication-with-social-accounts-and-brokering.adoc[leveloffset=+1]
 include::proc_using-protocol-based-providers.adoc[leveloffset=+1]
 include::proc_managing-users-using-keycloak.adoc[leveloffset=+1]
+include::proc_configuring-prod-short-to-use-external-keycloak.adoc[leveloffset=+1]
```
`prod-short` in file name; should be `che`

rkratky commented on src/main/pages/che-7/administration-guide/proc_configuring-prod-short-to-use-external-keycloak.adoc:
```diff
@@ -0,0 +1,77 @@
+// configuring-authorization
+
+[id="configuring-prod-short-to-use-external-keycloak_{context}"]
```
```suggestion
[id="configuring-{prod-short}-to-use-external-keycloak_{context}"]
```

amisevsk reviewed (commented):
Added minor fixes from my review, but not sure if PR is ready to merge as there are still some grey areas (and also we need to downstream new properties and make sure it's not mangled).

Commit:
Remove spurious context.

Add article in restr. env.

Add assembly for importing self-signed certs to allow direct link (#1493)

Create an assembly in quick starts to contain process for importing
self-signed certs. This allows it to be linked directly from the
Che server when there is a certificate error

Signed-off-by: Angel Misevski <amisevsk@redhat.com>

Fix downstr (#1499)

* Remove spurious context.

* Add assembly for importing self-signed certs to allow direct link (#1493)

Create an assembly in quick starts to contain process for importing
self-signed certs. This allows it to be linked directly from the
Che server when there is a certificate error

Signed-off-by: Angel Misevski <amisevsk@redhat.com>

* Add article in restr. env.

Co-authored-by: Angel Misevski <amisevsk@redhat.com>

Ext. Keycloak fixes.

First sketch of docs for using an External Keycloak

Signed-off-by: David Festal <dfestal@redhat.com>

Small fixes after looking at the preview

Signed-off-by: David Festal <dfestal@redhat.com>

Update proc_configuring-prod-short-to-use-external-keycloak.adoc

Update proc_configuring-prod-short-to-use-external-keycloak.adoc

Update src/main/pages/che-7/administration-guide/proc_configuring-prod-short-to-use-external-keycloak.adoc

Co-authored-by: Angel Misevski <amisevsk@redhat.com>

Update src/main/pages/che-7/administration-guide/proc_configuring-prod-short-to-use-external-keycloak.adoc

Co-authored-by: David Festal <dfestal@redhat.com>

Extract placeholders used in examples to table at start

Signed-off-by: Angel Misevski <amisevsk@redhat.com>

Extract identity provider docs links to properties for downstream

Signed-off-by: Angel Misevski <amisevsk@redhat.com>

Minor wording fix for external keycloak doc

Signed-off-by: Angel Misevski <amisevsk@redhat.com>

rkratky reviewed (approved):
Thanks, @davidfestal. I fixed a bunch of things and rebased to get a single commit.

If any of the discussions in this PR's comments need to lead to modifications, let's deal with that in future PRs. Merging now.
